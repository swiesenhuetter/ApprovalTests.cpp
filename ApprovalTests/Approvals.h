#pragma once

#include <string>
#include <functional>
#include <exception>

#include "ApprovalTests/ApprovalTestsVersion.h"
#include "ApprovalTests/core/FileApprover.h"
#include "reporters/DefaultReporter.h"
#include "reporters/DefaultReporterDisposer.h"
#include "ApprovalTests/core/Reporter.h"
#include "namers/ApprovalTestNamer.h"
#include "writers/ExistingFile.h"
#include "namers/ExistingFileNamer.h"
#include "namers/SubdirectoryDisposer.h"
#include "namers/DefaultNamerDisposer.h"
#include "scrubbers/Scrubbers.h"

namespace ApprovalTests
{

    class Approvals
    {
    private:
        Approvals() = default;

        ~Approvals() = default;

    public:
        static std::shared_ptr<ApprovalNamer> getDefaultNamer()
        {
            return DefaultNamerFactory::getDefaultNamer()();
        }

        static void verify(std::string contents,
                           const Reporter& reporter = DefaultReporter())
        {
            verifyScrubbed(contents, Scrubbers::doNothing, reporter);
        }

        static void verifyScrubbed(std::string contents,
                                   Scrubber scrubber,
                                   const Reporter& reporter = DefaultReporter())
        {
            verifyWithExtension(scrubber(contents), ".txt", reporter);
        }

        static void verifyWithExtension(std::string contents,
                                        const std::string& fileExtensionWithDot,
                                        const Reporter& reporter = DefaultReporter())
        {
            StringWriter writer(contents, fileExtensionWithDot);
            FileApprover::verify(*getDefaultNamer(), writer, reporter);
        }

        static void verify(const ApprovalWriter& writer,
                           const Reporter& reporter = DefaultReporter())
        {
            FileApprover::verify(*getDefaultNamer(), writer, reporter);
        }

        template <typename T>
        using IsNotDerivedFromWriter =
            typename std::enable_if<!std::is_base_of<ApprovalWriter, T>::value,
                                    int>::type;

        template <typename T, typename = IsNotDerivedFromWriter<T>>
        static void verify(const T& contents,
                           const Reporter& reporter = DefaultReporter())
        {
            verify(StringUtils::toString(contents), reporter);
        }

        template <typename T, typename = IsNotDerivedFromWriter<T>>
        static void verifyWithExtension(const T& contents,
                                        const std::string& fileExtensionWithDot,
                                        const Reporter& reporter = DefaultReporter())
        {
            verifyWithExtension(
                StringUtils::toString(contents), fileExtensionWithDot, reporter);
        }

        template <typename T,
                  typename Function,
                  typename = Detail::EnableIfNotDerivedFromReporter<Function>>
        static void verify(const T& contents,
                           Function converter,
                           const Reporter& reporter = DefaultReporter())
        {
            std::stringstream s;
            converter(contents, s);
            verify(s.str(), reporter);
        }

        template <typename T,
                  typename Function,
                  typename = Detail::EnableIfNotDerivedFromReporter<Function>>
        static void verifyWithExtension(const T& contents,
                                        Function converter,
                                        const std::string& fileExtensionWithDot,
                                        const Reporter& reporter = DefaultReporter())
        {
            std::stringstream s;
            converter(contents, s);
            verifyWithExtension(s.str(), fileExtensionWithDot, reporter);
        }

        static void verifyExceptionMessage(std::function<void(void)> functionThatThrows,
                                           const Reporter& reporter = DefaultReporter())
        {
            std::string message = "*** no exception thrown ***";
            try
            {
                functionThatThrows();
            }
            catch (const std::exception& e)
            {
                message = e.what();
            }
            verify(message, reporter);
        }

        template <typename Iterator>
        static void verifyAll(
            std::string header,
            const Iterator& start,
            const Iterator& finish,
            std::function<void(typename Iterator::value_type, std::ostream&)> converter,
            const Reporter& reporter = DefaultReporter())
        {
            std::stringstream s;
            if (!header.empty())
            {
                s << header << "\n\n\n";
            }
            for (auto it = start; it != finish; ++it)
            {
                converter(*it, s);
                s << '\n';
            }
            verify(s.str(), reporter);
        }

        template <typename Container>
        static void verifyAll(
            std::string header,
            const Container& list,
            std::function<void(typename Container::value_type, std::ostream&)> converter,
            const Reporter& reporter = DefaultReporter())
        {
            verifyAll<typename Container::const_iterator>(
                header, list.begin(), list.end(), converter, reporter);
        }

        template <typename T>
        static void verifyAll(std::string header,
                              const std::vector<T>& list,
                              const Reporter& reporter = DefaultReporter())
        {
            int i = 0;
            verifyAll<std::vector<T>>(
                header,
                list,
                [&](T e, std::ostream& s) { s << "[" << i++ << "] = " << e; },
                reporter);
        }

        template <typename T>
        static void verifyAll(const std::vector<T>& list,
                              const Reporter& reporter = DefaultReporter())
        {
            verifyAll<T>("", list, reporter);
        }

        static void verifyExistingFile(const std::string filePath,
                                       const Reporter& reporter = DefaultReporter())
        {
            ExistingFile writer(filePath);
            ExistingFileNamer namer(filePath);
            FileApprover::verify(namer, writer, reporter);
        }

        static SubdirectoryDisposer
        useApprovalsSubdirectory(std::string subdirectory = "approval_tests")
        {
            return SubdirectoryDisposer(subdirectory);
        }

        static DefaultReporterDisposer
        useAsDefaultReporter(const std::shared_ptr<Reporter>& reporter)
        {
            return DefaultReporterDisposer(reporter);
        }

        static FrontLoadedReporterDisposer
        useAsFrontLoadedReporter(const std::shared_ptr<Reporter>& reporter)
        {
            return FrontLoadedReporterDisposer(reporter);
        }

        static DefaultNamerDisposer useAsDefaultNamer(NamerCreator namerCreator)
        {
            return DefaultNamerDisposer(namerCreator);
        }
    };
}
