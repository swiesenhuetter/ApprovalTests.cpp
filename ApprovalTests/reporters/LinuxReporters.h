#pragma once

#include "DiffPrograms.h"
#include "GenericDiffReporter.h"
#include "FirstWorkingReporter.h"

namespace ApprovalTests
{
    namespace Linux
    {
        class SublimeMergeSnapReporter : public GenericDiffReporter
        {
        public:
            SublimeMergeSnapReporter()
                : GenericDiffReporter(DiffPrograms::Linux::SUBLIME_MERGE_SNAP())
            {
                launcher.setForeground(true);
            }
        };

        class SublimeMergeFlatpakReporter : public GenericDiffReporter
        {
        public:
            SublimeMergeFlatpakReporter()
                : GenericDiffReporter(DiffPrograms::Linux::SUBLIME_MERGE_FLATPAK())
            {
                launcher.setForeground(true);
            }
        };

        class SublimeMergeRepositoryPackageReporter : public GenericDiffReporter
        {
        public:
            SublimeMergeRepositoryPackageReporter()
                : GenericDiffReporter(
                      DiffPrograms::Linux::SUBLIME_MERGE_REPOSITORY_PACKAGE())
            {
                launcher.setForeground(true);
            }
        };

        class SublimeMergeDirectDownloadReporter : public GenericDiffReporter
        {
        public:
            SublimeMergeDirectDownloadReporter()
                : GenericDiffReporter(
                      DiffPrograms::Linux::SUBLIME_MERGE_DIRECT_DOWNLOAD())
            {
                launcher.setForeground(true);
            }
        };

        class SublimeMergeReporter : public FirstWorkingReporter
        {
        public:
            SublimeMergeReporter()
                : FirstWorkingReporter({new SublimeMergeSnapReporter(),
                                        new SublimeMergeFlatpakReporter(),
                                        new SublimeMergeRepositoryPackageReporter(),
                                        new SublimeMergeDirectDownloadReporter()})
            {
            }
        };

        class KDiff3Reporter : public GenericDiffReporter
        {
        public:
            KDiff3Reporter() : GenericDiffReporter(DiffPrograms::Linux::KDIFF3())
            {
            }
        };

        class MeldReporter : public GenericDiffReporter
        {
        public:
            MeldReporter() : GenericDiffReporter(DiffPrograms::Linux::MELD())
            {
            }
        };

        class BeyondCompareReporter : public GenericDiffReporter
        {
        public:
            BeyondCompareReporter()
                : GenericDiffReporter(DiffPrograms::Linux::BEYOND_COMPARE())
            {
            }
        };

        class LinuxDiffReporter : public FirstWorkingReporter
        {
        public:
            LinuxDiffReporter()
                : FirstWorkingReporter({
                      // begin-snippet: linux_diff_reporters
                      new BeyondCompareReporter(),
                      new MeldReporter(),
                      new SublimeMergeReporter(),
                      new KDiff3Reporter()
                      // end-snippet
                  })
            {
            }
        };
    }
}
