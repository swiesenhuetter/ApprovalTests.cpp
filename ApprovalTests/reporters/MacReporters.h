#pragma once

#include "DiffPrograms.h"
#include "GenericDiffReporter.h"
#include "FirstWorkingReporter.h"

namespace ApprovalTests
{
    namespace Mac
    {
        class DiffMergeReporter : public GenericDiffReporter
        {
        public:
            DiffMergeReporter() : GenericDiffReporter(DiffPrograms::Mac::DIFF_MERGE())
            {
            }
        };

        // begin-snippet: add_reporter_class
        class AraxisMergeReporter : public GenericDiffReporter
        {
        public:
            AraxisMergeReporter() : GenericDiffReporter(DiffPrograms::Mac::ARAXIS_MERGE())
            {
            }
        };
        // end-snippet

        class VisualStudioCodeReporter : public GenericDiffReporter
        {
        public:
            VisualStudioCodeReporter() : GenericDiffReporter(DiffPrograms::Mac::VS_CODE())
            {
            }
        };

        class BeyondCompareReporter : public GenericDiffReporter
        {
        public:
            BeyondCompareReporter()
                : GenericDiffReporter(DiffPrograms::Mac::BEYOND_COMPARE())
            {
            }
        };

        class KaleidoscopeReporter : public GenericDiffReporter
        {
        public:
            KaleidoscopeReporter()
                : GenericDiffReporter(DiffPrograms::Mac::KALEIDOSCOPE())
            {
            }
        };

        class SublimeMergeReporter : public GenericDiffReporter
        {
        public:
            SublimeMergeReporter()
                : GenericDiffReporter(DiffPrograms::Mac::SUBLIME_MERGE())
            {
                launcher.setForeground(true);
            }
        };

        class KDiff3Reporter : public GenericDiffReporter
        {
        public:
            KDiff3Reporter() : GenericDiffReporter(DiffPrograms::Mac::KDIFF3())
            {
            }
        };

        class P4MergeReporter : public GenericDiffReporter
        {
        public:
            P4MergeReporter() : GenericDiffReporter(DiffPrograms::Mac::P4MERGE())
            {
            }
        };

        class TkDiffReporter : public GenericDiffReporter
        {
        public:
            TkDiffReporter() : GenericDiffReporter(DiffPrograms::Mac::TK_DIFF())
            {
            }
        };

        class MacDiffReporter : public FirstWorkingReporter
        {
        public:
            MacDiffReporter()
                : FirstWorkingReporter({
                      // begin-snippet: mac_diff_reporters
                      new AraxisMergeReporter(),
                      new BeyondCompareReporter(),
                      new DiffMergeReporter(),
                      new KaleidoscopeReporter(),
                      new P4MergeReporter(),
                      new SublimeMergeReporter(),
                      new KDiff3Reporter(),
                      new TkDiffReporter(),
                      new VisualStudioCodeReporter()
                      // end-snippet
                  })
            {
            }
        };
    }
}
