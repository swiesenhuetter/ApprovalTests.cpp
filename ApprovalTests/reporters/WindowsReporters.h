#pragma once

#include "DiffPrograms.h"
#include "FirstWorkingReporter.h"
#include "GenericDiffReporter.h"

namespace ApprovalTests
{
    namespace Windows
    {

        class VisualStudioCodeReporter : public GenericDiffReporter
        {
        public:
            VisualStudioCodeReporter()
                : GenericDiffReporter(DiffPrograms::Windows::VS_CODE())
            {
            }
        };

        // ----------------------- Beyond Compare ----------------------------------
        class BeyondCompare3Reporter : public GenericDiffReporter
        {
        public:
            BeyondCompare3Reporter()
                : GenericDiffReporter(DiffPrograms::Windows::BEYOND_COMPARE_3())
            {
            }
        };

        class BeyondCompare4Reporter : public GenericDiffReporter
        {
        public:
            BeyondCompare4Reporter()
                : GenericDiffReporter(DiffPrograms::Windows::BEYOND_COMPARE_4())
            {
            }
        };

        class BeyondCompareReporter : public FirstWorkingReporter
        {
        public:
            BeyondCompareReporter()
                : FirstWorkingReporter(
                      {new BeyondCompare4Reporter(), new BeyondCompare3Reporter()})
            {
            }
        };

        // ----------------------- Tortoise SVN ------------------------------------
        class TortoiseImageDiffReporter : public GenericDiffReporter
        {
        public:
            TortoiseImageDiffReporter()
                : GenericDiffReporter(DiffPrograms::Windows::TORTOISE_IMAGE_DIFF())
            {
            }
        };

        class TortoiseTextDiffReporter : public GenericDiffReporter
        {
        public:
            TortoiseTextDiffReporter()
                : GenericDiffReporter(DiffPrograms::Windows::TORTOISE_TEXT_DIFF())
            {
            }
        };

        class TortoiseDiffReporter : public FirstWorkingReporter
        {
        public:
            TortoiseDiffReporter()
                : FirstWorkingReporter(
                      {new TortoiseTextDiffReporter(), new TortoiseImageDiffReporter()})
            {
            }
        };

        // ----------------------- Tortoise Git ------------------------------------
        class TortoiseGitTextDiffReporter : public GenericDiffReporter
        {
        public:
            TortoiseGitTextDiffReporter()
                : GenericDiffReporter(DiffPrograms::Windows::TORTOISE_GIT_TEXT_DIFF())
            {
            }
        };

        class TortoiseGitImageDiffReporter : public GenericDiffReporter
        {
        public:
            TortoiseGitImageDiffReporter()
                : GenericDiffReporter(DiffPrograms::Windows::TORTOISE_GIT_IMAGE_DIFF())
            {
            }
        };

        class TortoiseGitDiffReporter : public FirstWorkingReporter
        {
        public:
            TortoiseGitDiffReporter()
                : FirstWorkingReporter({new TortoiseGitTextDiffReporter(),
                                        new TortoiseGitImageDiffReporter()})
            {
            }
        };

        // -------------------------------------------------------------------------
        class WinMergeReporter : public GenericDiffReporter
        {
        public:
            WinMergeReporter()
                : GenericDiffReporter(DiffPrograms::Windows::WIN_MERGE_REPORTER())
            {
            }
        };

        class AraxisMergeReporter : public GenericDiffReporter
        {
        public:
            AraxisMergeReporter()
                : GenericDiffReporter(DiffPrograms::Windows::ARAXIS_MERGE())
            {
            }
        };

        class CodeCompareReporter : public GenericDiffReporter
        {
        public:
            CodeCompareReporter()
                : GenericDiffReporter(DiffPrograms::Windows::CODE_COMPARE())
            {
            }
        };

        class SublimeMergeReporter : public GenericDiffReporter
        {
        public:
            SublimeMergeReporter()
                : GenericDiffReporter(DiffPrograms::Windows::SUBLIME_MERGE())
            {
                launcher.setForeground(true);
            }
        };

        class KDiff3Reporter : public GenericDiffReporter
        {
        public:
            KDiff3Reporter() : GenericDiffReporter(DiffPrograms::Windows::KDIFF3())
            {
            }
        };

        class WindowsDiffReporter : public FirstWorkingReporter
        {
        public:
            WindowsDiffReporter()
                : FirstWorkingReporter({
                      // begin-snippet: windows_diff_reporters
                      new TortoiseDiffReporter(), // Note that this uses Tortoise SVN Diff
                      new TortoiseGitDiffReporter(),
                      new BeyondCompareReporter(),
                      new WinMergeReporter(),
                      new AraxisMergeReporter(),
                      new CodeCompareReporter(),
                      new SublimeMergeReporter(),
                      new KDiff3Reporter(),
                      new VisualStudioCodeReporter(),
                      // end-snippet
                  })
            {
            }
        };
    }
}
