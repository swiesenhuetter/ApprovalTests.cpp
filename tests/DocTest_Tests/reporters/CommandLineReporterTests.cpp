#include "doctest/doctest.h"
#include "ApprovalTests/Approvals.h"
#include "ApprovalTests/reporters/WindowsReporters.h"
#include "ApprovalTests/reporters/MacReporters.h"
#include "ApprovalTests/reporters/LinuxReporters.h"

#include <vector>

using namespace ApprovalTests;

std::string scrubProgramFiles(const std::string& output)
{
    auto scrubbed = output;
    for (const auto& path : DiffInfo::getProgramFileLocations())
    {
        scrubbed =
            StringUtils::replaceAll(scrubbed, path, DiffInfo::programFileTemplate());
    }
    return scrubbed;
}

TEST_CASE("Test Command Lines")
{
    std::stringstream stream;
    std::vector<std::shared_ptr<GenericDiffReporter>> reporters = {
        // begin-snippet: add_reporter_command_test
        // Mac
        std::make_shared<Mac::AraxisMergeReporter>(),
        std::make_shared<Mac::BeyondCompareReporter>(),
        std::make_shared<Mac::DiffMergeReporter>(),
        std::make_shared<Mac::KaleidoscopeReporter>(),
        std::make_shared<Mac::P4MergeReporter>(),
        std::make_shared<Mac::SublimeMergeReporter>(),
        std::make_shared<Mac::KDiff3Reporter>(),
        std::make_shared<Mac::TkDiffReporter>(),
        std::make_shared<Mac::VisualStudioCodeReporter>(),
        // end-snippet

        // Windows
        std::make_shared<Windows::TortoiseTextDiffReporter>(),
        std::make_shared<Windows::TortoiseImageDiffReporter>(),
        std::make_shared<Windows::TortoiseGitTextDiffReporter>(),
        std::make_shared<Windows::TortoiseGitImageDiffReporter>(),
        std::make_shared<Windows::BeyondCompare3Reporter>(),
        std::make_shared<Windows::BeyondCompare4Reporter>(),
        std::make_shared<Windows::WinMergeReporter>(),
        std::make_shared<Windows::AraxisMergeReporter>(),
        std::make_shared<Windows::CodeCompareReporter>(),
        std::make_shared<Windows::SublimeMergeReporter>(),
        std::make_shared<Windows::KDiff3Reporter>(),
        std::make_shared<Windows::VisualStudioCodeReporter>(),

        // Linux
        std::make_shared<Linux::BeyondCompareReporter>(),
        std::make_shared<Linux::MeldReporter>(),
        std::make_shared<Linux::SublimeMergeSnapReporter>(),
        std::make_shared<Linux::SublimeMergeFlatpakReporter>(),
        std::make_shared<Linux::SublimeMergeRepositoryPackageReporter>(),
        std::make_shared<Linux::SublimeMergeDirectDownloadReporter>(),
        std::make_shared<Linux::KDiff3Reporter>()};
    for (const auto& reporter : reporters)
    {
        reporter->useCygwinConversions(false);
        reporter->launcher.invokeForWindows(true);
        stream << "windows: " << reporter->getCommandLine("a.txt", "b.txt") << '\n';

        reporter->launcher.invokeForWindows(false);
        stream << "unix   : " << reporter->getCommandLine("a.txt", "b.txt") << '\n';

        reporter->useCygwinConversions(true);
        reporter->launcher.invokeForWindows(false);
        stream << "cygwin : " << reporter->getCommandLine("a.txt", "b.txt") << '\n';
        stream << '\n';
    }
    Approvals::verify(scrubProgramFiles(stream.str()));
}
