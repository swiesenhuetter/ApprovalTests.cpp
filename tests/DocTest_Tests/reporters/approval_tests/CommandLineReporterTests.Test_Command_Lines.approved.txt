windows: start "" "/Applications/Araxis Merge.app/Contents/Utilities/compare" "a.txt" "b.txt"
unix   : "/Applications/Araxis Merge.app/Contents/Utilities/compare" "a.txt" "b.txt" &
cygwin : "$(cygpath '/Applications/Araxis Merge.app/Contents/Utilities/compare')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: start "" "/Applications/Beyond Compare.app/Contents/MacOS/bcomp" "a.txt" "b.txt"
unix   : "/Applications/Beyond Compare.app/Contents/MacOS/bcomp" "a.txt" "b.txt" &
cygwin : "$(cygpath '/Applications/Beyond Compare.app/Contents/MacOS/bcomp')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: start "" "/Applications/DiffMerge.app/Contents/MacOS/DiffMerge" "a.txt" "b.txt" -nosplash
unix   : "/Applications/DiffMerge.app/Contents/MacOS/DiffMerge" "a.txt" "b.txt" -nosplash &
cygwin : "$(cygpath '/Applications/DiffMerge.app/Contents/MacOS/DiffMerge')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" -nosplash &

windows: start "" "/Applications/Kaleidoscope.app/Contents/MacOS/ksdiff" "a.txt" "b.txt"
unix   : "/Applications/Kaleidoscope.app/Contents/MacOS/ksdiff" "a.txt" "b.txt" &
cygwin : "$(cygpath '/Applications/Kaleidoscope.app/Contents/MacOS/ksdiff')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: start "" "/Applications/p4merge.app/Contents/MacOS/p4merge" "a.txt" "b.txt"
unix   : "/Applications/p4merge.app/Contents/MacOS/p4merge" "a.txt" "b.txt" &
cygwin : "$(cygpath '/Applications/p4merge.app/Contents/MacOS/p4merge')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: cmd /S /C ""/Applications/Sublime Merge.app/Contents/SharedSupport/bin/smerge" mergetool --no-wait "a.txt" "b.txt" -o "b.txt""
unix   : "/Applications/Sublime Merge.app/Contents/SharedSupport/bin/smerge" mergetool --no-wait "a.txt" "b.txt" -o "b.txt"
cygwin : "$(cygpath '/Applications/Sublime Merge.app/Contents/SharedSupport/bin/smerge')" mergetool --no-wait "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" -o "$(cygpath -aw 'b.txt')"

windows: start "" "/Applications/kdiff3.app/Contents/MacOS/kdiff3" "a.txt" "b.txt" -m
unix   : "/Applications/kdiff3.app/Contents/MacOS/kdiff3" "a.txt" "b.txt" -m &
cygwin : "$(cygpath '/Applications/kdiff3.app/Contents/MacOS/kdiff3')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" -m &

windows: start "" "/Applications/TkDiff.app/Contents/MacOS/tkdiff" "a.txt" "b.txt"
unix   : "/Applications/TkDiff.app/Contents/MacOS/tkdiff" "a.txt" "b.txt" &
cygwin : "$(cygpath '/Applications/TkDiff.app/Contents/MacOS/tkdiff')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: start "" "/Applications/Visual Studio Code.app/Contents/Resources/app/bin/code" -d "a.txt" "b.txt"
unix   : "/Applications/Visual Studio Code.app/Contents/Resources/app/bin/code" -d "a.txt" "b.txt" &
cygwin : "$(cygpath '/Applications/Visual Studio Code.app/Contents/Resources/app/bin/code')" -d "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: start "" "{ProgramFiles}TortoiseSVN\bin\TortoiseMerge.exe" "a.txt" "b.txt"
unix   : "{ProgramFiles}TortoiseSVN\bin\TortoiseMerge.exe" "a.txt" "b.txt" &
cygwin : "$(cygpath '{ProgramFiles}TortoiseSVN\bin\TortoiseMerge.exe')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: start "" "{ProgramFiles}TortoiseSVN\bin\TortoiseIDiff.exe" /left:"a.txt" /right:"b.txt"
unix   : "{ProgramFiles}TortoiseSVN\bin\TortoiseIDiff.exe" /left:"a.txt" /right:"b.txt" &
cygwin : "$(cygpath '{ProgramFiles}TortoiseSVN\bin\TortoiseIDiff.exe')" /left:"$(cygpath -aw 'a.txt')" /right:"$(cygpath -aw 'b.txt')" &

windows: start "" "{ProgramFiles}TortoiseGit\bin\TortoiseGitMerge.exe" "a.txt" "b.txt"
unix   : "{ProgramFiles}TortoiseGit\bin\TortoiseGitMerge.exe" "a.txt" "b.txt" &
cygwin : "$(cygpath '{ProgramFiles}TortoiseGit\bin\TortoiseGitMerge.exe')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: start "" "{ProgramFiles}TortoiseGit\bin\TortoiseGitIDiff.exe" /left:"a.txt" /right:"b.txt"
unix   : "{ProgramFiles}TortoiseGit\bin\TortoiseGitIDiff.exe" /left:"a.txt" /right:"b.txt" &
cygwin : "$(cygpath '{ProgramFiles}TortoiseGit\bin\TortoiseGitIDiff.exe')" /left:"$(cygpath -aw 'a.txt')" /right:"$(cygpath -aw 'b.txt')" &

windows: start "" "{ProgramFiles}Beyond Compare 3\BCompare.exe" "a.txt" "b.txt"
unix   : "{ProgramFiles}Beyond Compare 3\BCompare.exe" "a.txt" "b.txt" &
cygwin : "$(cygpath '{ProgramFiles}Beyond Compare 3\BCompare.exe')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: start "" "{ProgramFiles}Beyond Compare 4\BCompare.exe" "a.txt" "b.txt"
unix   : "{ProgramFiles}Beyond Compare 4\BCompare.exe" "a.txt" "b.txt" &
cygwin : "$(cygpath '{ProgramFiles}Beyond Compare 4\BCompare.exe')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: start "" "{ProgramFiles}WinMerge\WinMergeU.exe" "a.txt" "b.txt"
unix   : "{ProgramFiles}WinMerge\WinMergeU.exe" "a.txt" "b.txt" &
cygwin : "$(cygpath '{ProgramFiles}WinMerge\WinMergeU.exe')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: start "" "{ProgramFiles}Araxis\Araxis Merge\Compare.exe" "a.txt" "b.txt"
unix   : "{ProgramFiles}Araxis\Araxis Merge\Compare.exe" "a.txt" "b.txt" &
cygwin : "$(cygpath '{ProgramFiles}Araxis\Araxis Merge\Compare.exe')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: start "" "{ProgramFiles}Devart\Code Compare\CodeCompare.exe" "a.txt" "b.txt"
unix   : "{ProgramFiles}Devart\Code Compare\CodeCompare.exe" "a.txt" "b.txt" &
cygwin : "$(cygpath '{ProgramFiles}Devart\Code Compare\CodeCompare.exe')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: cmd /S /C ""{ProgramFiles}Sublime Merge\smerge.exe" mergetool --no-wait "a.txt" "b.txt" -o "b.txt""
unix   : "{ProgramFiles}Sublime Merge\smerge.exe" mergetool --no-wait "a.txt" "b.txt" -o "b.txt"
cygwin : "$(cygpath '{ProgramFiles}Sublime Merge\smerge.exe')" mergetool --no-wait "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" -o "$(cygpath -aw 'b.txt')"

windows: start "" "{ProgramFiles}KDiff3\kdiff3.exe" "a.txt" "b.txt"
unix   : "{ProgramFiles}KDiff3\kdiff3.exe" "a.txt" "b.txt" &
cygwin : "$(cygpath '{ProgramFiles}KDiff3\kdiff3.exe')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: start "" "{ProgramFiles}Microsoft VS Code\Code.exe" -d "a.txt" "b.txt"
unix   : "{ProgramFiles}Microsoft VS Code\Code.exe" -d "a.txt" "b.txt" &
cygwin : "$(cygpath '{ProgramFiles}Microsoft VS Code\Code.exe')" -d "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: start "" "bcompare" "a.txt" "b.txt"
unix   : "bcompare" "a.txt" "b.txt" &
cygwin : "$(cygpath 'bcompare')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: start "" "meld" "a.txt" "b.txt"
unix   : "meld" "a.txt" "b.txt" &
cygwin : "$(cygpath 'meld')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &

windows: cmd /S /C ""/snap/bin/sublime-merge" mergetool --no-wait "a.txt" "b.txt" -o "b.txt""
unix   : "/snap/bin/sublime-merge" mergetool --no-wait "a.txt" "b.txt" -o "b.txt"
cygwin : "$(cygpath '/snap/bin/sublime-merge')" mergetool --no-wait "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" -o "$(cygpath -aw 'b.txt')"

windows: cmd /S /C ""/var/lib/flatpak/exports/bin/com.sublimemerge.App" mergetool --no-wait "a.txt" "b.txt" -o "b.txt""
unix   : "/var/lib/flatpak/exports/bin/com.sublimemerge.App" mergetool --no-wait "a.txt" "b.txt" -o "b.txt"
cygwin : "$(cygpath '/var/lib/flatpak/exports/bin/com.sublimemerge.App')" mergetool --no-wait "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" -o "$(cygpath -aw 'b.txt')"

windows: cmd /S /C ""smerge" mergetool --no-wait "a.txt" "b.txt" -o "b.txt""
unix   : "smerge" mergetool --no-wait "a.txt" "b.txt" -o "b.txt"
cygwin : "$(cygpath 'smerge')" mergetool --no-wait "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" -o "$(cygpath -aw 'b.txt')"

windows: cmd /S /C ""/opt/sublime_merge/sublime_merge" mergetool --no-wait "a.txt" "b.txt" -o "b.txt""
unix   : "/opt/sublime_merge/sublime_merge" mergetool --no-wait "a.txt" "b.txt" -o "b.txt"
cygwin : "$(cygpath '/opt/sublime_merge/sublime_merge')" mergetool --no-wait "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" -o "$(cygpath -aw 'b.txt')"

windows: start "" "kdiff3" "a.txt" "b.txt"
unix   : "kdiff3" "a.txt" "b.txt" &
cygwin : "$(cygpath 'kdiff3')" "$(cygpath -aw 'a.txt')" "$(cygpath -aw 'b.txt')" &


